[4.0] Update buttons.php layout to latest Bootstrap 5 markup, also added LEGEND markup

thednp opened the pull request:
### Summary of Changes
Updated layout file according to Bootstrap 5 new markup
* the `active` class no longer required for the `<label>` element
* the `<input>` is no longer inside the `<label>` element
* the new markup requires no `bootstrap.Button` plugin functionality anymore, everything is pure CSS

Details on the updated markup
https://getbootstrap.com/docs/5.0/components/button-group/#checkbox-and-radio-button-groups

Pull Request for Issues #32366 and #32275.

### Testing Instructions
* Open `site/templates/cassiopeia/templateDetails.xml`
* Create a new field inside the `<fieldset name="advanced">`, here a quick example

```xml
<field
  name="my-radio"
  type="Radio"
  default="0"
  class="btn-group"
  label="Radio Test">
  <option value="0">Disabled</option>
  <option value="1">Enabled</option>
</field>
```
* Login to `site/administrator` 
* go to `System` -> `Site Template Styles` -> `Cassiopeia`
* Check the new `Radio Test`, it should work perfect

### Actual result BEFORE applying this Pull Request
The layout won't work with the new Bootstrap 5 CSS and Markup. The JavaScript for radio / checkboxes toggling has been deprecated in favor of a more traditional markup and adjusted CSS.

### Expected result AFTER applying this Pull Request
If instructed as above. or in any case of a `Radio` field used with `class="btn-group"` attribute, the radio input works properly in the sense that showcases the correct chosen option on any user selection.

### Notice
Due to the radio/checkboxes toggle functionality being deprecated in Bootstrap 5 (but replaced with a new CSS + Markup solution) we might need to check for any redundant JavaScript related.

### Documentation Changes Required
None

thednp committed:
Update buttons.php

Updated layout file according to Bootstrap 5 new markup
* the `active` class no longer required for the label
* the `<input>` is no longer inside the `<label>` element
* the new markup requires no `bootstrap.Button` plugin functionality anymore

Details on the updated markup
https://getbootstrap.com/docs/5.0/components/button-group/#checkbox-and-radio-button-groups

Quy commented:
I have tested this item :white_check_mark: successfully on c53b0c66ac0fd40ed1c43015e53a4c67d7bbbe62<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

OctavianC commented:
I have tested this item :white_check_mark: successfully on c53b0c66ac0fd40ed1c43015e53a4c67d7bbbe62<br /><br />It fixes #32366 and #32275<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

infograf768 commented:
Agree this solves the original issue but here is the difference between using bootstrap button and switcher:

<img width="478" alt="Screen Shot 2021-02-10 at 09 58 05" src="https://user-images.githubusercontent.com/869724/107487598-a1d7db80-6b86-11eb-9f98-9cb17d5b5909.png">

There is no background color difference between Yes and No.
Also, is this totally a11y compliant?

thednp commented:
@infograf768 regarding the `Sticky Header` option you have

here's how it should work
![yesno](https://user-images.githubusercontent.com/6761246/107491367-f37e6700-6b82-11eb-9c36-97ddab214747.png)


In regards to a11y, I don't think there are any issues when compared to previous version.

thednp commented:
@infograf768 I will provide a fix for the layout to work with `btn-group-yesno` asap, a small bug I found.

infograf768 commented:
Adding indeed `btn-group-yesno` is much better

<img width="178" alt="Screen Shot 2021-02-10 at 10 39 48" src="https://user-images.githubusercontent.com/869724/107492117-545e6d00-6b8c-11eb-91a2-a035f3426688.png">

infograf768 commented:
@thednp 
You should also consider adding a legend for a11y, as we have for the switcher.

<img width="591" alt="Screen Shot 2021-02-10 at 10 34 01" src="https://user-images.githubusercontent.com/869724/107492414-b3bc7d00-6b8c-11eb-8451-1610b921fabb.png">

thednp commented:
Roger that, stay close for the updated layout.

thednp commented:
Alright, I've changed in my dev so that this layout will do the following:
* if you use `class="btn-group"` for any `type="Radio"` field, it should work as expected, which means the container of the input options will use `class="btn-group"`, no change here
* however, if you use `class="btn-group-yesno"` for `type="Radio"`, it will also use `class="btn-group"`, because `class="btn-group-yesno"` does not exist

Now about a11y, @infograf768 I don't see why a visible `<label>` need to have a legend for our `buttons` layout here.

thednp committed:
Update buttons.php

Regarding https://github.com/joomla/joomla-cms/pull/32367#issuecomment-776590252

thednp commented:
@infograf768 the switcher layout indeed is a special case, it needs to tell screen readers what's the option about since there is no visible `<label>`, only a hidden `<legend>` which I'm sure readers check.

You think we need to add `<legend>` for `type="Radio"` fields when using this very `buttons` layout?

thednp committed:
Update buttons.php

LoL typo

richard67 commented:
> You think we need to add `<legend>` for `type="Radio"` fields when using this very `buttons` layout?

Maybe @chmst can have a look on that, too.

infograf768 commented:
If our class remains `class="btn-group btn-group-yesno"
Your last changes creates a double btn-group class

<img width="919" alt="Screen Shot 2021-02-10 at 11 11 32" src="https://user-images.githubusercontent.com/869724/107495758-c933a600-6b90-11eb-974d-890bf358e4ce.png">

Therefore the class would have to be `class="btn-group-yesno"` which is NOT bc as J3 uses `class="btn-group btn-group-yesno"

Concerning the legend, we should do the same as for switcher which has a visible label

<img width="645" alt="Screen Shot 2021-02-10 at 11 13 09" src="https://user-images.githubusercontent.com/869724/107496027-19ab0380-6b91-11eb-94df-629f2ed59ba9.png">

richard67 commented:
I'm also not happy with the last change, that looks like a hack to me, that str_replace.

chmst commented:
A legend is necessary (visually hidden).

thednp commented:
@chmst I will add `<legend>` then, no problem.

@richard67 and @infograf768 you don't have to use `class="btn-group btn-group-yesno"`, just `class="btn-group-yesno"`, but if you REALLY think that's the way to do it, no problem, change incoming.

infograf768 commented:
> but if you REALLY think that's the way to do it, no problem, change incoming.

It would be B/C with J3. As we have already so much non B/C in j4, let's not add one. ;)

richard67 commented:
> @richard67 and @infograf768 you don't have to use `class="btn-group btn-group-yesno"`, just `class="btn-group-yesno"`, but if you REALLY think that's the way to do it, no problem, change incoming.

@thednp Your str_replace would not be safe for the case someone uses a custom class like e.g. `my-fancy-class-yesno` for some reason. You should have done a regex replace to make sure you only replace what you want to replace ;-) But for B/C we should revert that completely, like @infograf768 said.

thednp committed:
Update buttons.php

Changes: 
* Reverting to previous as explained [here](https://github.com/joomla/joomla-cms/pull/32367#issuecomment-776610851), now `class="btn-group btn-group-yesno"` is necessary to create a switcher with `type="Radio"` field with the `buttons` layout
* added `<legend>` as mentioned [here](https://github.com/joomla/joomla-cms/pull/32367#issuecomment-776606460)
* both `.switcher__legend` and the new `radio__legend` additional styling is irrelevant since both elements use `.visually-hidden`, the `.switcher__legend` is now redundant

thednp commented:
@chmst @infograf768 and @richard67 done.

See commit note.

infograf768 commented:
The `switcher__legend` class exists in css, but not the `radio__legend`

thednp commented:
Yes, the `.radio__legend` has just been created. But I was referring to the fact that the CSS for `.switcher__legend` is redundant, unless it serves a purpose for J3?

infograf768 commented:
imho, `switcher_legend` css as well as `radio-legend` are totally useless as the only thing `switcher_legend` does is setting the font-size for a visually-hidden legend.
But let's @drmenzelit comment on that.

ceford commented:
When I click the button it changes from disabled to enabled. But the two hidden checkboxes don't change checked status. I find this whole thing very confusing!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

thednp commented:
@ceford it's the same [here](https://getbootstrap.com/docs/5.0/components/button-group/#checkbox-and-radio-button-groups), but it's OK, you can check the `checked` status via JavaScript, here's how:

* open Inspector (assume you use a Chrome based browser)
* right-click on an option, then select `Inspect`
* in the browser console, type in `$0.checked`, it should return the current state

It's how the browser engine works now, it won't update the `checked` attribute, but internally it always knows the current state, as it renders your `.btn` elements properly according to the real `checked` state.

infograf768 commented:
Hmm, I now get
<img width="200" alt="Screen Shot 2021-02-10 at 12 29 04" src="https://user-images.githubusercontent.com/869724/107504484-a8bd1900-6b9b-11eb-8ad3-5a0d4aa4cb67.png">

richard67 commented:
> Hmm, I now get ...

@infograf768 I guess you mean that space between the "No" and "Yes" buttons, right?

thednp commented:
@infograf768 @richard67 that's what my previous fix was about. my implementation wasn't perfect as Rich pointed out, but would have solved that. So yea, you need to use `class="btn-group btn-group-yesno"` to have it working without spacing between.

infograf768 commented:
Forget it
Typo in the xml
I had `class="btn-goup btn-group-yesno"`

instead of
`class="btn-group btn-group-yesno"`

all ok

<img width="191" alt="Screen Shot 2021-02-10 at 12 48 44" src="https://user-images.githubusercontent.com/869724/107506435-5c270d00-6b9e-11eb-9db1-457d941b6960.png">

sorry

richard67 commented:
I have tested this item :white_check_mark: successfully on 13fe5a0db0890bcb406847493d8fb59d65c65e22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

OctavianC commented:
I have tested this item :white_check_mark: successfully on 13fe5a0db0890bcb406847493d8fb59d65c65e22<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

brianteeman commented:
Not sure why this is RTC as there are outstanding issues. There is no point in adding a class to the legend that doesn't exist in the css. It serves no purpose as the legend is never visually displayed. Or do I miss something?

richard67 commented:
> Not sure why this is RTC as there are outstanding issues. There is no point in adding a class to the legend that doesn't exist in the css. It serves no purpose as the legend is never visually displayed. Or do I miss something?

We have such classes according to BEM elsewhere, too, as once added by @ciar4n , as far as I remember.

brianteeman commented:
on something that is hidden? That makes zero sense.

richard67 commented:
@thednp Could you remove the `radio__legend ` from the class of the legend? Thanks in advance.

richard67 commented:
Back to pending.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

thednp committed:
Update buttons.php

thednp commented:
> 
> 
> @thednp Could you remove the `radio__legend ` from the class of the legend? Thanks in advance.

Done.

richard67 commented:
I have tested this item :white_check_mark: successfully on 9aabd48da16217f6ae3737fd076f5985f12a35ef<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

infograf768 commented:
I have tested this item :white_check_mark: successfully on 9aabd48da16217f6ae3737fd076f5985f12a35ef<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

infograf768 commented:
rtc
let’s get rid of switcher__legend code and css in another pr<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32367">issues.joomla.org/tracker/joomla-cms/32367</a>.</sub>

thednp commented:
@infograf768 I would do it rn if I knew where's the source.

richard67 commented:
@thednp `find ./ -type f -exec grep -Hn "switcher__legend" {} \;` in a Linux shell or `findstr /s /c:"switcher__legend" *.*` in a Windows Command window will help you to find out when executed in the Joomla root folder.

richard67 commented:
But for SCSS sources in the build folder it needs a git clone of course, that's not included in a normal installation package.

thednp commented:
Added https://github.com/joomla/joomla-cms/pull/32377 in regards to #issuecomment-776805109

thednp commented:
Boy I wish we had the same involvement for #30816

infograf768 commented:
Tks
